### PR TITLE
Add producers engine

### PR DESCRIPTION
**Pull request by sauloperez**
Still missing:

- [x] Study whether to use ENUM at db level for unit
- [ ] Add some meaningful seed data

### CODE REVIEW

File: engines/account/lib/account/engine.rb [hunk 1]
```diff
@@ -1,14 +1,9 @@
+require_relative '../../../../lib/engines/with_migrations'
+
 module Account
   class Engine < ::Rails::Engine
+    extend ::Engines::WithMigrations
 
     isolate_namespace Account
-
-    initializer 'account.append_migrations' do |app|
```
Comment (sauloperez) [inline engines/account/lib/account/engine.rb]: I DRYied this to the module above

**Comment by sauloperez** [comment]
I still need to remove a lot of crab from the engine

File: lib/engines/with_migrations.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module Engines
+  # It DRYies the common snippet to append engine's migrations to the main app
+  module WithMigrations
```
Comment (enricostano) [inline lib/engines/with_migrations.rb]: put this in `Shared` engine please, is where all the common stuff goes. Makes sense?

File: engines/producers/MIT-LICENSE [hunk 3]
```diff
@@ -0,0 +1,20 @@
+Copyright 2015 YOURNAME
```
Comment (enricostano) [inline engines/producers/MIT-LICENSE]: MEC

File: engines/producers/Rakefile [hunk 4]
```diff
@@ -0,0 +1,34 @@
+begin
```
Comment (enricostano) [inline engines/producers/Rakefile]: If we don't have specific rake tasks maybe we can just remove this file and the tasks directory from engines

File: engines/producers/db/migrate/20150815081517_create_providers.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class CreateProviders < ActiveRecord::Migration
+  def change
+    create_table :providers do |t|
+      t.string :name
+      t.string :email
+      t.text :address
```
Comment (enricostano) [inline engines/producers/db/migrate/20150815081517_create_providers.rb]: All required I guess, no?

File: engines/producers/db/migrate/20150815082649_create_products.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class CreateProducts < ActiveRecord::Migration
+  def change
+    create_table :products do |t|
+      t.string :name
+      t.integer :price
+      t.string :unit
+      t.references :provider, index: true
```
Comment (enricostano) [inline engines/producers/db/migrate/20150815082649_create_products.rb]: All required I guess, no?

File: engines/producers/app/models/producers/product.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+module Producers
+  class Product < ActiveRecord::Base
+    belongs_to :provider
```
Comment (enricostano) [inline engines/producers/app/models/producers/product.rb]: needs to validate at least `unit` to be included in supported units, right?

File: lib/engines/with_migrations.rb (hunk 2 above)
Comment (sauloperez) [inline lib/engines/with_migrations.rb]: I tried it but AFAIK it won't be easy. We'll have to either force load `shared` before other engines or  add shared as dependency. I tried the latter with no luck.

I can still require the lib from `shared` but, then what is the responsibility of the main app?

File: engines/producers/lib/tasks/producers_tasks.rake [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# desc "Explaining what the task does"
```
Comment (enricostano) [inline engines/producers/lib/tasks/producers_tasks.rake]: :fire:

File: engines/account/lib/account/engine.rb (hunk 1 above)
Comment (enricostano) [inline engines/account/lib/account/engine.rb]: `Shared` engine is the place I guess

File: lib/engines/with_migrations.rb (hunk 2 above)
Comment (enricostano) [inline lib/engines/with_migrations.rb]: Mh leave it in `lib`for now then, I'll investigate. But thanks for the abstraction! :clap:

**Comment by enricostano** [comment]
Thanks for the effort @sauloperez !!! :clap: :clap: :clap: 
Please add yourself in https://github.com/coopdevs/katuma/blob/develop/public/humans.txt :stuck_out_tongue_winking_eye:

File: engines/producers/db/migrate/20150815081517_create_providers.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class CreateProviders < ActiveRecord::Migration
+  def change
+    create_table :providers do |t|
```
Comment (enricostano) [inline engines/producers/db/migrate/20150815081517_create_providers.rb]: This should be `producers` right?

File: engines/producers/app/models/producers/provider.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+module Producers
+  class Provider < ActiveRecord::Base
```
Comment (enricostano) [inline engines/producers/app/models/producers/provider.rb]: This should be `Producer` right?

File: engines/producers/app/models/producers/product.rb (hunk 7 above)
Comment (sauloperez) [inline engines/producers/app/models/producers/product.rb]: Yep, but also I'd like to use some sort of enum

File: engines/producers/app/views/layouts/producers/application.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  <%= stylesheet_link_tag    "producers/application", media: "all" %>
+  <%= javascript_include_tag "producers/application" %>
+  <%= csrf_meta_tags %>
+</head>
+<body>
+
+<%= yield %>
+
+</body>
+</html>
```
Comment (andresgutgon) [inline engines/producers/app/views/layouts/producers/application.html.erb]: :fire: No need more base layouts

File: engines/producers/app/assets/javascripts/producers/application.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+// Any JavaScript/Coffee file within this directory, lib/assets/javascripts, vendor/assets/javascripts,
+// or vendor/assets/javascripts of plugins, if any, can be referenced here using a relative path.
+//
+// It's not advisable to add code directly here, but if you do, it'll appear at the bottom of the
+// compiled file.
+//
+// Read Sprockets README (https://github.com/sstephenson/sprockets#sprockets-directives) for details
+// about supported directives.
+//
+//= require_tree .
```
Comment (andresgutgon) [inline engines/producers/app/assets/javascripts/producers/application.js]: :fire: No needed for now. Lets avoid this boilerplate files. Thanks Rails but not needed for now

File: engines/producers/app/assets/stylesheets/producers/application.css [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+ * or vendor/assets/stylesheets of plugins, if any, can be referenced here using a relative path.
+ *
+ * You're free to add application-wide styles to this file and they'll appear at the bottom of the
+ * compiled file so the styles you add here take precedence over styles defined in any styles
+ * defined in the other CSS/SCSS files in this directory. It is generally better to create a new
+ * file per style scope.
+ *
+ *= require_tree .
+ *= require_self
+ */
```
Comment (andresgutgon) [inline engines/producers/app/assets/stylesheets/producers/application.css]: Ditto boilerplate

File: engines/producers/app/views/layouts/producers/application.html.erb (hunk 11 above)
Comment (andresgutgon) [inline engines/producers/app/views/layouts/producers/application.html.erb]: I was thinking about how to organize UI interface. Maybe we could create an interface engine and put all visuals there (rails view/controllers, javascripts, css,...)

And leave the rest of engines as services used in Interface engine controllers... I not sure splitting UI in engines is a good aproach.

Let's discuss this offline  in some moment.

/cc @enricostano @sauloperez

File: engines/producers/app/assets/javascripts/producers/application.js (hunk 12 above)
Comment (sauloperez) [inline engines/producers/app/assets/javascripts/producers/application.js]: lol

File: engines/producers/app/models/producers/product.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+module Producers
+  class Product < ActiveRecord::Base
+    self.table_name = :products
+
+    belongs_to :provider
+
+    enum unit: [:kg, :pc, :lt]
```
Comment (enricostano) [inline engines/producers/app/models/producers/product.rb]: can we enforce it at DB level too?

File: engines/producers/db/migrate/20150815082649_create_products.rb [hunk 15]
```diff
@@ -0,0 +1,12 @@
+class CreateProducts < ActiveRecord::Migration
+  def change
+    create_table :products do |t|
+      t.string :name, null: false
+      t.integer :price, null: false
+      t.string :unit, default: 0, null: false
```
Comment (enricostano) [inline engines/producers/db/migrate/20150815082649_create_products.rb]: ditto

File: engines/producers/app/models/producers/product.rb (hunk 14 above)
Comment (andresgutgon) [inline engines/producers/app/models/producers/product.rb]: maybe better names: `kilogrames`, ...

File: engines/producers/db/migrate/20150815082649_create_products.rb (hunk 15 above)
Comment (enricostano) [inline engines/producers/db/migrate/20150815082649_create_products.rb]: BTW, this should be integer, right? Not stringa :trollface: 

http://api.rubyonrails.org/v4.1.0/classes/ActiveRecord/Enum.html

**Comment by sauloperez** [comment]
This is ready to be merged. We can add seed data in a future PR.

File: lib/engines/with_migrations.rb (hunk 2 above)
Comment (sauloperez) [inline lib/engines/with_migrations.rb]: It turns out this is the way to go with migrations in CBRA :point_right:  http://blog.pivotal.io/pivotal-labs/labs/leave-your-migrations-in-your-rails-engines.

I think I'll extract it as a gem and add it as a dependency. What do you think? I'll probably do the same in Redbooth, btw.

File: public/humans.txt [hunk 16]
```diff
@@ -7,6 +7,11 @@
 	Twitter: @enricostano
 	From: Barcelona, Catalonia, Spain
 
+	Backend: Pau Pérez
```
Comment (andresgutgon) [inline public/humans.txt]: :muscle:

**Comment by andresgutgon** [comment]
:+1:

File: Gemfile [hunk 17]
```diff
@@ -16,15 +16,17 @@ gem 'unicorn', '4.9.0'
...
-gemspec path: "engines/shared"
-gemspec path: "engines/account"
-gemspec path: "engines/landing"
-gemspec path: "engines/group"
+gemspec path: 'engines/shared'
+gemspec path: 'engines/account'
+gemspec path: 'engines/landing'
+gemspec path: 'engines/group'
+gemspec path: 'engines/producers'
+gemspec path: 'engines/suppliers'
```
Comment (enricostano) [inline Gemfile]: suppliers too? seems wrong to me

File: engines/producers/app/models/producers/producer.rb [hunk 18]
```diff
@@ -0,0 +1,7 @@
+module Producers
+  class Producer < ActiveRecord::Base
+    self.table_name = :producers
+
+    has_many :products
```
Comment (enricostano) [inline engines/producers/app/models/producers/producer.rb]: no validations at AR level? there are 3 `null: false` in the DB

File: db/schema.rb [hunk 19]
```diff
@@ -73,6 +73,25 @@
...
+    t.string   "email",      null: false
+    t.text     "address",    null: false
+    t.datetime "created_at"
+    t.datetime "updated_at"
+  end
+
+  create_table "products", force: true do |t|
+    t.string   "name",                    null: false
+    t.integer  "price",                   null: false
+    t.integer  "unit",        default: 0, null: false
```
Comment (enricostano) [inline db/schema.rb]: not sure we need a default here, `0` is for `kg` here, looking at AR model

File: lib/engines/with_migrations.rb (hunk 2 above)
Comment (enricostano) [inline lib/engines/with_migrations.rb]: Yep I agree, but we can do it in another PR if you want

**Comment by enricostano** [comment]
:mailbox_with_no_mail: :+1:

File: engines/producers/app/models/producers/product.rb (hunk 14 above)
Comment (sauloperez) [inline engines/producers/app/models/producers/product.rb]: Wait, I'll need to specify the DB int too

File: Gemfile (hunk 17 above)
Comment (andresgutgon) [inline Gemfile]: :fire:

File: public/humans.txt [hunk 20]
```diff
@@ -7,6 +7,11 @@
...
+	Contact: saulopefa [at] gmail.com
+	Twitter: @pau_prez
```
Comment (enricostano) [inline public/humans.txt]: `@prez_pau` :trollface:

Comment (sauloperez) [inline public/humans.txt]: loooool
